### PR TITLE
Created ssh.add_authorized_key function - Closes #3665

### LINKED ISSUE #3665: Create helper to upload ssh key to remote server
Create helper similar that we have for ssh module `upload_file`, but
1) Check that remote folder (.ssh) exists and create if not
2) Check whether file `authorized_keys` exists and create if not
3) Append content (from `~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub` for example) to `authorized_keys` file (probably through `echo 'key' >> ~/.ssh/authorized_keys`)

### PR DESCRIPTION
**Pull request by rochacbruno**
Closes #3665

```console
$ cd robottelo
$ python manage.py shell
In [1]: from robottelo.ssh  import add_authorized_key
In [2]: add_authorized_key('/tmp/test_rsa.pub')  # path, file-like or string-key

2016-07-28 20:06:12 - robottelo.ssh - INFO - Instantiated Paramiko client 0x7fd1f97fb690
2016-07-28 20:06:12 - robottelo.ssh - DEBUG - Connected to [FQDN.redhat.com]
2016-07-28 20:06:12 - robottelo.ssh - DEBUG - >>> mkdir -p ~/.ssh
2016-07-28 20:06:13 - robottelo.ssh - DEBUG - >>> grep -q 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDuMCPmX7iBXAxH5oLznswA5ccfV/FABwIWnYl0OYRkDhv3mu9Eogk4H6sCguq4deJtRkwg2C3yEmsNYfBWYu4y5RkI4TH/k3N161wn91nBxs/+wqoN3g9tUuWrf98PG4NnYvmZU67RuiSUNXpgLEPfo8jMKkJ5veKu++DmHdpfqFB9ljWEfWz+kAAKgwo251VRDaKwFsb91LbLFpqn9rfMJUBhOn+Uebfd0TrHzw08gbVmvfAn61isvFVhvIJBTjNSWsBIm8SuCvhH+inYOwttfE8FGeR1KSp9Xl0PCYDK0BwvQO3qwD+nehsEUR/FJUXm1IZPc8fi17ieGgPOnrgf user@new-host' ~/.ssh/authorized_keys || echo 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDuMCPmX7iBXAxH5oLznswA5ccfV/FABwIWnYl0OYRkDhv3mu9Eogk4H6sCguq4deJtRkwg2C3yEmsNYfBWYu4y5RkI4TH/k3N161wn91nBxs/+wqoN3g9tUuWrf98PG4NnYvmZU67RuiSUNXpgLEPfo8jMKkJ5veKu++DmHdpfqFB9ljWEfWz+kAAKgwo251VRDaKwFsb91LbLFpqn9rfMJUBhOn+Uebfd0TrHzw08gbVmvfAn61isvFVhvIJBTjNSWsBIm8SuCvhH+inYOwttfE8FGeR1KSp9Xl0PCYDK0BwvQO3qwD+nehsEUR/FJUXm1IZPc8fi17ieGgPOnrgf user@new-host' >> ~/.ssh/authorized_keys
2016-07-28 20:06:13 - robottelo.ssh - DEBUG - >>> chmod 700 ~/.ssh
2016-07-28 20:06:14 - robottelo.ssh - DEBUG - >>> chmod 600 ~/.ssh/authorized_keys
2016-07-28 20:06:15 - robottelo.ssh - DEBUG - >>> chown -R root ~/.ssh
2016-07-28 20:06:16 - robottelo.ssh - DEBUG - >>> command -v restorecon && restorecon -RvF ~/.ssh || true
2016-07-28 20:06:16 - robottelo.ssh - DEBUG - <<< stdout /usr/sbin/restorecon
2016-07-28 20:06:16 - robottelo.ssh - INFO - Destroying Paramiko client 0x7fd1f97fb690
2016-07-28 20:06:17 - robottelo.ssh - INFO - Destroyed Paramiko client 0x7fd1f97fb690
```

### CODE REVIEW

**Comment by oshtaier** [comment]
Wow. Very nice PR Bruno. Huge thanks to you that you implemented it so fast

One more thing and I am really sorry that I forgot to mention it in my initial issue.
Current PR will require one improvement. It is very good to support uploading of ssh key from local path, but I will need also possibility to copy key from remote server, so basically I need:
1) Take key from satellite server (like qe-rhel6 machine)
2) Put it on remote VM host (that I created and started)
like append qe-rhel6>~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub 1235332.server.com>~/.ssh/authorized_keys

Once more, sorry about that

**Comment by rochacbruno** [comment]
@oshtaier since robottelo doen't run on the Sat machine, you will need to do that in 2 steps.
 
```python
from robottelo import ssh

# get key from satellite box
# without passing hostname it gets from settings.server
foreman_key = ssh.command('cat ~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub ').stdout

# put key to the client host
ssh.add_authorized_key(foreman_key, hostname='235332.server.com', username=..., password=....)
``` 

We can transform those 2 steps in a new helper, what do you think? Or do you want to execute the cross-ssh command in the satellite box?

**Comment by oshtaier** [comment]
Maybe we can still run two helpers consequently. So we can just add small helper as a reader for the key:
`foreman_key = ssh.command('cat ~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub ').stdout`
or we can create a new helper that can read both from local directory and remote one.
Anyway, helper approach seems more attractive for me

File: robottelo/helpers.py [hunk 1]
```diff
@@ -253,3 +253,20 @@ def remove_katello_ca(hostname=None):
             )
             if result.return_code != 0:
                 raise AssertionError('Failed to reset the rhsm.conf')
+
+
+def add_remote_execution_ssh_key(hostname, key_path=None, **kwargs):
```
Comment (rochacbruno) [inline robottelo/helpers.py]: @oshtaier added `robottelo.helpers.add_remote_execution_ssh_key('123455.redhat,com')` so this helpers does both things 1 ) get key from sat machine 2) add to the client machine.

I am not sure on how the machine running the test will connect to the client box, is it using username and password? or some default key?

BTW: that **kwargs passes username, password, key to the client connection

Comment (oshtaier) [inline robottelo/helpers.py]: So basically, we should use Jenkins key to connect to both servers. There is a chance that we use Beaker password to connect to Sat machine and Jenkins key to VM, but I don't think that is supported at that moment

**Comment by abalakh** [comment]
The code was looking good for me so i decided to clone it and test it out... and it worked as a charm :) Was able to copy ssh key, and to successfully use remote execution feature. And i've tried connection by both login/password and login/ssh-key - both ways it worked ok. Great job :+1: 
I do have one comment though - regardless of whether the `authorized_keys` file (and `.ssh` folder) was present or not, you're trying to update permissions and playing with selinux, which imo isn't the best option. Think about use case when user wants to add 10 ssh keys to `authorized_keys` file. In such case permissions, owner and selinux will be updated 10 times with no particular reason for that. I'd like propose to update permissions/owner/selinux only in case `.ssh` folder and `authorized_keys` aren't in place, and in opposite case it looks like user has his environment all set, so if there's a problem with permissions - it's user's responsibility, we shouldn't touch anything. But that's discussable.
Apart from that - great PR, ACK :+1:

File: robottelo/ssh.py [hunk 2]
```diff
@@ -97,13 +106,72 @@ def _get_connection(hostname=None, username=None, password=None,
...
+    :param int timeout: Time to wait for establish the connection.
+    """
+
+    if getattr(key, 'read', None):  # key is a file-like object
+        key_content = key.read()
+    elif is_ssh_pub_key(key):  # key is a valid key-string
+        key_content = key
+    elif os.path.exists(key):  # key is a path to a pub key-file
+        with open(key, 'r') as key_file:
+            key_content = key_file.read()
```
Comment (sghai) [inline robottelo/ssh.py]: Nice to see we are covering all possible inputs.

**Comment by sghai** [comment]
PR looks great.

**Comment by oshtaier** [comment]
ACK

**Comment by abalakh** [comment]
As per IRC discussion:
>brunorocha> oshtaier: abalakht: if we choose to check permissions only when file does not exists we need to add 2 more ssh calls

>brunorocha> oshtaier: abalakht: the remote_ssh provisioning template does that (update permissions on every key added)

> brunorocha> abalakht: oshtaier: So I think we can merge it if you are good with the PR.. in future if changing permissions become a problem we cam change that.

**Pull request merged**